integr8ly/integreatly-operator
# Improve multitenancy user creation/deletion

# Issue link
https://issues.redhat.com/browse/MGDAPI-3074

# What
1. We want to only create tenants for users that have logged in after installation of RHOAM, not all of users
2. We want to create tenants based on User CRs, not identity CRs
3. We want to trigger deletion of tenants based on missing user CR, not identity CR
4. Removal of the code around reading only identities related to a given by envar IDP name (since we have moved to user CRs and are no longer dependant on IDP)
5. Add e2e test for multitenant RHOAM

# Verification steps
1. Provision OSD cluster
2. Download this repo https://github.com/codeready-toolchain/toolchain-e2e/tree/master/setup and from the repo run:
- `export QUAY_NAMESPACE=<YOUR_QUAY_NS>`
- `make dev-deploy-e2e`
- `go run setup/main.go --users 2 --default 2 -b 2 --custom 0 --username test1`
3. From this branch run: 
- `INSTALLATION_TYPE=multitenant-managed-api make cluster/prepare/local`
- `IDENTITY_PROVIDER_NAME=rhd INSTALLATION_TYPE=multitenant-managed-api make code/run`
4. Wait for install to complete and confirm that no new routes under 3scale ns are present (for test1 users, only our regular routes should be present)
5. From toolchain repo run:
- `go run setup/main.go --users 2 --default 2 -b 2 --custom 0 --username test2`
6. Wait 3 minutes and confirm that 3scale routes for 2 new tenants have been created, confirm by navigating to the admin route and confirm that you can click on login via SSO.
7. Remove UserSignUp CR for one of the users of test2
8. Wait few minutes and confirm that the admin route for the user you have removed is no longer available (alternatively, login as 3scale master and confirm that the tenant is scheduled for deletion)
9. To test e2e test either expect multitenant-rhoam-e2e to pass on this pr (we have an issue with our e2e, so might be better to run the e2e locally against fresh install [no idp setup or toolchain operator] and run the single test)
